core/src/main/scala/org/http4s/Message.scala

```diff
@@ -76,44 +74,153 @@ sealed trait Message[F[_]] extends MessageOps[F] { self =>
+        body.through(util.decode(cs))
+    }
+
+  // General header methods
 
-  def contentType: Option[`Content-Type`] = headers.get(`Content-Type`)
+  def transformHeaders(f: Headers => Headers): Self =
+    change(headers = f(headers))
 
-  /** Returns the charset parameter of the `Content-Type` header, if present. Does
-    * not introspect the body for media types that define a charset
-    * internally.
+  /** Keep headers that satisfy the predicate
+    *
+    * @param f predicate
+    * @return a new message object which has only headers that satisfy the predicate
     */
-  def charset: Option[Charset] = contentType.flatMap(_.charset)
+  final def filterHeaders(f: Header => Boolean): Self =
+    transformHeaders(_.filter(f))
 
-  def isChunked: Boolean =
-    headers.get(`Transfer-Encoding`).exists(_.values.contains_(TransferCoding.chunked))
+  final def removeHeader(key: HeaderKey): Self =
+    filterHeaders(_.isNot(key))
+
+  /** Add the provided headers to the existing headers, replacing those of the same header name
+    * The passed headers are assumed to contain no duplicate Singleton headers.
+    */
+  final def putHeaders(headers: Header*): Self =
+    transformHeaders(_.put(headers: _*))
+
+  /** Replace the existing headers with those provided */
+  @deprecated("Use withHeaders instead", "0.20.0-M2")
+  final def replaceAllHeaders(headers: Headers): Self =
+    withHeaders(headers)
+
+  /** Replace the existing headers with those provided */
+  @deprecated("Use withHeaders instead", "0.20.0-M2")
+  final def replaceAllHeaders(headers: Header*): Self =
+    withHeaders(headers: _*)
+
+  final def withTrailerHeaders(trailerHeaders: F[Headers]): Self =
+    withAttribute(Message.Keys.TrailerHeaders[F], trailerHeaders)
+
+  final def withoutTrailerHeaders: Self =
+    withoutAttribute(Message.Keys.TrailerHeaders[F])
 
   /**
-    * The trailer headers, as specified in Section 3.6.1 of RFC 2616.  The resulting
-    * task might not complete unless the entire body has been consumed.
+    * The trailer headers, as specified in Section 3.6.1 of RFC 2616. The resulting
+    * F might not complete unless the entire body has been consumed.
     */
   def trailerHeaders(implicit F: Applicative[F]): F[Headers] =
     attributes.get(Message.Keys.TrailerHeaders[F]).getOrElse(F.pure(Headers.empty))
 
+  // Specific header methods
+
+  @deprecated("Use withContentType(`Content-Type`(t)) instead", "0.20.0-M2")
+  final def withType(t: MediaType)(implicit F: Functor[F]): Self =
+    withContentType(`Content-Type`(t))
+
+  final def withContentType(contentType: `Content-Type`): Self =
+    putHeaders(contentType)
+
+  final def withoutContentType: Self =
+    filterHeaders(_.isNot(`Content-Type`))
+
+  @deprecated("Use Option.fold instead", "0.20.0-M2")
+  final def withContentTypeOption(contentTypeO: Option[`Content-Type`]): Self =
+    contentTypeO.fold(withoutContentType)(withContentType)
+
+  lazy val contentType: Option[`Content-Type`] =
```

Comment:
Are we sure that laziness pays off?  It adds weight to every request and response, and only is a win if read multiple times.  We've moved away from a lot of our prior use of lazies.

Furthermore, the risk of contention is low.  We can make these faster with the null init pattern, though it's a bit more verbose.

Reply:
No clue. If we decide against the laziness, then I just think we should be consistent and have defs for all of the things. And yeah, most of these are going to be read at most once.

Reply:
Wow, there's a lot more laziness than I thought.  I thought you were reintroducing more.

That said, I think it's (almost?) always a mistake.  For comparison's sake, I just checked out akka-http's `HttpMessage` and found no lazies.